api/src/services/authentication.ts

```diff
@@ -409,17 +411,55 @@ export class AuthenticationService {
 			issuer: 'directus',
 		});
 
-		await this.knex('directus_sessions')
-			.update({
-				token: newRefreshToken,
-				expires: refreshTokenExpiration,
-			})
-			.where({ token: refreshToken });
+		if (options?.session) {
+			if (!record.session_next) {
+				// keep the old session active for a second
+				const SAFETY_TIMEOUT = 1000;
+
+				// deal with a sliding window
+				await this.knex('directus_sessions').insert({
+					token: newRefreshToken,
+					user: record.user_id,
+					expires: refreshTokenExpiration,
+					ip: this.accountability?.ip,
+					user_agent: this.accountability?.userAgent,
+					origin: this.accountability?.origin,
+				});
+
+				await this.knex('directus_sessions')
+					.update({
+						next_token: newRefreshToken,
+						expires: new Date(Date.now() + SAFETY_TIMEOUT),
+					})
+					.where({ token: refreshToken });
+			} else {
+				// update the next session expiry date
+				await this.knex('directus_sessions')
+					.update({
+						expires: refreshTokenExpiration,
+					})
+					.where({ token: newRefreshToken });
+			}
+		} else {
+			// original stateless token behavior
+			await this.knex('directus_sessions')
+				.update({
+					token: newRefreshToken,
+					expires: refreshTokenExpiration,
+				})
+				.where({ token: refreshToken });
+		}
 
 		if (record.user_id) {
 			await this.knex('directus_users').update({ last_access: new Date() }).where({ id: record.user_id });
 		}
 
+		// clear expired sessions for the current user
+		await this.knex('directus_sessions')
+			.delete()
+			.where('user', '=', record.user_id)
+			.andWhere('expires', '<', new Date());
```

Comment:
Will this cause issues for large projects?

Reply:
Do you mean performance- or timeout-wise?

Reply:
I was thinking performance wise as i recall an issue where the clearing of sessions took a long time on a big instance since we only do that on login. I tried to limit that issue as much as possible by explicitly looking for the current users sessions not all of them.